hotshot-query-service/src/data_source/fetching/state_cert.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the HotShot Query Service library.
+//
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not,
+// see <https://www.gnu.org/licenses/>.
+
+//! Fetching for light client state update certificates.
+
+use std::{future::IntoFuture, sync::Arc};
+
+use anyhow::bail;
+use async_trait::async_trait;
+use derive_more::From;
+use futures::future::FutureExt;
+use hotshot_types::traits::node_implementation::{ConsensusTime, NodeType};
+
+use super::Storable;
+use crate::{
+    availability::{QueryablePayload, StateCertQueryData},
+    data_source::{
+        fetching::{
+            AvailabilityProvider, FetchRequest, Fetchable, Fetcher, Notifiers, PassiveFetch,
+        },
+        storage::{
+            pruning::PrunedHeightStorage, AvailabilityStorage, NodeStorage,
+            UpdateAvailabilityStorage,
+        },
+        VersionedDataSource,
+    },
+    Payload, QueryError, QueryResult,
+};
+
+#[derive(Clone, Copy, Debug, From)]
+pub(super) struct StateCertRequest(u64);
+
+impl FetchRequest for StateCertRequest {}
+
+#[async_trait]
+impl<Types> Fetchable<Types> for StateCertQueryData<Types>
+where
+    Types: NodeType,
+    Payload<Types>: QueryablePayload<Types>,
+{
+    type Request = StateCertRequest;
+
+    /// Does this object satisfy the given request?
+    fn satisfies(&self, req: Self::Request) -> bool {
+        self.0.epoch.u64() == req.0
+    }
+
+    async fn active_fetch<S, P>(
+        tx: &mut impl AvailabilityStorage<Types>,
+        fetcher: Arc<Fetcher<Types, S, P>>,
+        req: Self::Request,
+    ) -> anyhow::Result<()>
+    where
+        S: VersionedDataSource + 'static,
+        for<'a> S::Transaction<'a>: UpdateAvailabilityStorage<Types>,
+        for<'a> S::ReadOnly<'a>:
+            AvailabilityStorage<Types> + NodeStorage<Types> + PrunedHeightStorage,
+        P: AvailabilityProvider<Types>,
+    {
+        match tx.get_state_cert(req.0).await {
+            Ok(state_cert) => {
+                fetcher.store_and_notify(state_cert).await;
+            },
+            Err(QueryError::NotFound | QueryError::Missing) => {
+                tracing::debug!("state cert not found; trying fetch");
+            },
+            Err(QueryError::Error { message }) => {
+                bail!("failed to get state cert for epoch {}: {message}", req.0);
+            },
+        }
```

Comment:
This logic does not make sense. It is saying, if we can already get the state cert from the database (via `tx`) then store it in the database and notify the caller. This is redundant, if this function is being called we should assume the cert is _not_ in the database.

Then, if it wasn't in the database (`NotFound` or `Missing`) we do nothing. So it will never be fetched from a peer. If we want to enable catchup/fetching of state certs from peers, this needs to do something like `fetcher.state_cert_fetcher.spawn_fetch(req)`

Reply:
Got it. If we don't want to implement peer fetching at this point, should we just return nothing for `FetchingDataSource::get_state_cert()`

Reply:
The top level `get_state_cert` has to return something if you want the query to work. But if you're fine with it failing whenever we don't have the state cert locally, rather than trying to get it from peers, you can have `active_fetch` just return `Ok(())` (see e.g. `Fetchable<Types> for TransactionQueryData<Types>` which also doesn't implement fetching)